### ISSUE TITLE
Building test

### ISSUE THREAD

**Issue by OmarD-ctrl** [ultralytics/ultralytics#3628]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi everyone,
Does anyone know how to make custom test on a test set? The val() is not enough for my dataset and am trying to make BaseTester and DetectionTester but I found problems with the dataloader.
I need to know the test accuracy and thus make test() on the test set
Thanks!

**Comment by jiapei100** [comment]
Same here... 
I wanna build **ultralytics** from source as well, without depending upon the **pip python-opencv**.

**Comment by glenn-jocher** [comment]
@OmarD-ctrl hi there,

To perform a custom test on a test set using YOLOv8, you can use the `test()` function. The `test()` function allows you to evaluate the accuracy of your trained model on a separate test set.

To use `test()`, you need to define a test dataset using a custom dataloader. This dataloader should load your test data and labels in a format that YOLOv8 can understand. Once you have the test dataset and dataloader ready, you can pass them to the `test()` function, which will evaluate your model on the test set and provide you with the test accuracy.

Make sure to follow the YOLOv8 documentation for more details on how to set up the dataloader and use the `test()` function correctly.

I hope this helps! Let me know if you have any further questions.

**Comment by OmarD-ctrl** [comment]
@glenn-jocher Hi, thank you for your response.

Well in fact, I've already prepared a test set and I've specified its path in the yaml file.
The problem is that YOLO does not litterally have a `test()` attribute. It has `predict()` but that's just an output it's not related to testing.

Please give me updates if I missed anything.
Thanks!

**Comment by glenn-jocher** [comment]
@jiapei100 hi there,

I understand your concern regarding testing with YOLOv8. YOLOv8 does not have a specific `test()` attribute like `predict()`. However, you can still perform testing by using the `val()` function.

In the context of YOLOv8, the `val()` function is used for both validation and testing. When you specify the path to your test set in the YAML file, YOLOv8 will automatically load the test data during the validation phase. This means that the `val()` function evaluates the model's performance on the test set as well.

Regarding the `LoadImagesAndLabels` class, it is used for loading images and labels for both training and validation, not testing. The `autosplit()` function, on the other hand, is responsible for splitting a dataset into train/val/test sets.

To summarize, you can perform testing with YOLOv8 by using the `val()` function with your test set specified in the YAML file. The `LoadImagesAndLabels` class is used for loading images and labels for training and validation, while `autosplit()` is used for splitting the dataset into different sets.

I hope this clarifies things for you. Let me know if you have any further questions.

**Comment by OmarD-ctrl** [comment]
@glenn-jocher hi, thank you for the response.

I clearly understand what you mean because that's what I thought in the beginning otherwise the test path in the yaml file would be meaningless. But this what made me conused: when I apply model.val() it starts printing the progress like this
Ultralytics YOLOv8.0.132 🚀 Python-3.10.12 torch-2.0.1+cu118 CUDA:0 (Tesla T4, 15102MiB)
YOLOv8n summary (fused): 168 layers, 3006233 parameters, 0 gradients
Downloading https://ultralytics.com/assets/Arial.ttf to /root/.config/Ultralytics/Arial.ttf...
100%|██████████| 755k/755k [00:00<00:00, 126MB/s]
val: Scanning /......./val/labels.cache...
the problem is the val line, it just precises that it's working on val dir images and labels whereas the test dir is not mentioned.

**Comment by glenn-jocher** [comment]
@OmarD-ctrl hi there,

I understand your concern about the confusion between the `val()` function and the test set in YOLOv8. When you use the `val()` function, it prints progress updates mentioning the validation directory, but it doesn't mention the test directory. This led you to wonder if the `val()` function is actually working on the test set.

In this case, if you want to manually test your model on the test set, you can modify the path used by the `val()` function to point to your test set. By doing this, you can ensure that the `val()` function works on your test set instead of the validation set.

Remember, YOLOv8 treats the test set as a separate entity during the validation phase. So, modifying the path in the `val()` function to reflect your test set will allow you to evaluate your model's performance on the test set directly.

I hope this solution helps! Let me know if you have any further questions.

**Comment by CiceStella** [comment]
Hi, I am facing the same problem, but reading the answers now I am confused because I understand that by having it available, the test set is also evaluated at each epoch which I didn't think was usually done. But maybe I misunderstood the previous answer: so wanting with a yolov8 model to test the performance on the duly labelled test set to evaluate PR metrics,  I have to use predict? Or Val? Thanks!

**Comment by glenn-jocher** [comment]
@CiceStella hi there!

I understand your confusion. With YOLOv8, there isn't a clear differentiation between validation and testing in the code. However, by general convention, we often use the validation set during the training of the model to aid in hyperparameter tuning and to gauge model performance without causing information bleed from the test set.

That said, using YOLOv8, if you want to evaluate the performance of your model on a separate, labeled test set, you would use the `val()` function similar to how you would for the validation set. Point this function to your test set, and it should give you the precision-recall metrics you're seeking. 

The `predict()` function, on the other hand, is generally used for generating model predictions on unlabelled data.

I hope that helps clear things up. If you have further questions, please feel free to ask!

**Comment by CiceStella** [comment]
Hello again, and sorry to bother you again. Let me summarise my situation: we have a network trained on 5 possible output values. For training we split the dataset into training, validation and test, created the yaml file 
and launched the training with:

from ultralytics import YOLO
model = YOLO('yolov8n.pt')  
model.train(data='../dataset/yolov8/labelimg/dataset/data.yaml', epochs=100, imgsz=1280)

the yaml file
train:[...]Scripts/dataset/yolov8/labelimg/dataset/train/images
val: [...]Scripts/dataset/yolov8/labelimg/dataset/valid/images
test: [...]Scripts/dataset/yolov8/labelimg/dataset/test/images
nc: 5
names: ['plate', 'green', 'yellow', 'red', 'orange']
so I expect that as usual the traning has been done ONLY on train and val set. (there is in fact no mention of the test data in the results).
Now I would like to try the network thus trained exclusively on the test set to get clean PR data. I tried as you said in the last post:
model.val(data='testImg.yaml', imgsz=1280) 
where tesImg.yaml is basically the same as prim
train: []/Scripts/dataset/yolov8/labelimg/dataset/train/images
val: []Scripts/dataset/yolov8/labelimg/dataset/valid/images
test: []Scripts/dataset/yolov8/labelimg/dataset/test/images
nc: 5
names: ['plate', 'green', 'yellow', 'red', 'orange']
 
(without putting train and val gives error)
But it does not produce acceptable results... Am I doing something wrong?
Thanks

**Comment by glenn-jocher** [comment]
@CiceStella Hi there! No bother at all; we're here to help!

With ultralytics YOLOv8 implementation, the "model.train()" function performs the training operation over the data specified in the 'train' key in the yaml file. Similarly, the "model.val()" function performs validation/testing on the data specified under the 'val' key in the yaml file only. This means that the 'test' key you're specifying in your yaml file is currently not being utilized in either training or validation/testing.

So, in order to test your model on the test set, you need to modify the 'val' key in a separate yaml file (which you've named as 'testImg.yaml') to point to your test images. This is what may have caused confusion in the previous conversation about using the 'val()' function for the test set. In this context, the 'val()' function is being repurposed to validate/test your network on the test set.

Here's a quick summary of the changes: 

1. trainImg.yaml (for training):
```
train: .../train/images
val: .../valid/images
nc: ...
names: ...
```

2. testImg.yaml (for testing):
```
(train key can be left as it is or same as in the trainImg.yaml)
val: .../test/images
nc: ...
names: ...
```

Now, when you run "model.train()" with 'trainImg.yaml' and "model.val()" with 'testImg.yaml', the model does not use the test data while training, and you can evaluate your model's performance cleanly on the test data.

I hope this clears things up! Feel free to reach out if you have any more questions or need further clarifications.
